src/org/recompile/mobile/PlatformPlayer.java

```diff
@@ -324,7 +324,12 @@ public long getMediaTime()
 
 		public boolean isRunning()
 		{
-			return wavClip.isRunning();
+			if(wavClip == null)

```

Comment:
Did you get my other comment?  I had thought I was writing a review, but I wasn't.  My draft with a stupid mistake at the beginning might have been sent out.  I corrected it, of course, but it might have been missed.  Anyhow, I explain how games play audio and why this check will probably be unnecessary once we fix the root of the problem.  I'll dig it up if you can't find it, otherwise, here's the relevant documentation:

https://docs.oracle.com/javame/config/cldc/ref-impl/midp2.0/jsr118/javax/microedition/media/Manager.html#createPlayer(java.io.InputStream,%20java.lang.String)

Reply:
Yes, got the other comment alongside those new ones... so, to keep things short, and make sure i'm getting things right: We do need to treat those cases where the game loads a NULL stream, but not by constantly checking if the clip is NULL, but instead by checking the stream right at the Player's constructor and throwing a IllegalArgumentException as needed? I'll try doing that, while also implementing the different string check recommended above (i'll apply it to the midi files as well, that kind of change doesn't seem like it will break things since we already expect those types to be lowercase).

Reply:
Yes.  According to the docs, the `createPlayer` methods are supposed to throw an `IllegalArgumentException` when passed a null stream.  I liked the idea of doing that in the constructor, but there isn't any reason that it couldn't also be done by an explicit check in `createPlayer` method.  There are merits to both, though we'd need to declare that thrown in the `PlatformPlayer` constructor, so it might ultimately be easier to just check for a null stream in `createPlayer` and throw there. There's nothing wrong with failing faster, after all.

Anyhow, the whole point is that there won't be a reason to check for nulls in `isRunning` because it won't be possible for there to be a null there.

Reply:
Okay, so... turns out the problem is different from what i was expecting. It's not that those games i tested are loading NULL streams. They are, in fact, loading files with the "audio/wav" type but either encoding those wav files differently (or else they would be interchangeable with "audio/x-wav"), or describing those files erroneously since i found out that the stream gets passed around correctly until the wavPlayer's constructor fails to open the Input Stream, which is behind these two hypothesis. 

Due to this, the "audio/wav" check was removed from the list and FreeJ2ME now assigns the stub "audioplayer" to that type. It might render a few games unable to play sound by using that specific wav type, but i'd rather leave them without audio than cause crashes on others. There's always the possibility of implementing more players for those different types later.

And well, a bit off topic but... thanks for being patient with the mess i've been doing on those commits and pull requests so far, i really appreciate it.

Reply:
Removing audio/wav is very likely to break other games.  (IIRC, audio/wav was the first type I supported, though I'd rather not go hunting for whatever game it was that I found that used it.  It was a long time ago and this is just a hobby project, after all.)  That's the thing about a project like this. 

The next step here is to examine the game to see why the wav file in question seems wrong.  Jar files are just zip files, so changing the extension from .jar to zip will let you unpack it.  From there, we can look at the asset to see if we can figure out why it's different, if that can be detected or corrected, etc.  If you've never used a hex editor, you're in for a treat.  

Oh, what game is it that appears to have the bad audio/wav asset?  It might help if I'm looking at the same thing you are.

We might also need to take a look at the games code to see what it's actually doing.  Fortunately, it's easy to find good-quality Java decompilers.  There are even some online, just upload a jar and get back decompiled source.   It can very helpful when diagnosing a particular game.  

Remember that this isn't a race.  Take your time and enjoy the puzzle solving part.  Just remember that when you're working on a game, it's not about getting the game to work so much as it is getting the API closer to the standard the game was written against.  (Also, I may periodically disappear for days or weeks.  I'm actually ridiculously busy most of the time, so I can't always be as quick to respond as I have been this weekend.)

Try to remember that everyone writing j2me games was testing on multiple conforming implementations.  For that reason, the javadocs are sacred scripture. Deviating from them just to get a game loaded will come back to bite you eventually.

Reply:
Hmm... apparently that was "audio/x-wav"  (I just took a look at an older release).  Ignore that bit.

Still, I think it's worth trying to work out why/how audio/wav that format differs from audio/x-wav.

Reply:
I'm also interested in finding out why those games (the one i'm testing for now is Assassin's Creed: Revelations, the 320x240 version) report a wav mime type but send files that cause trouble to the wavPlayer. I'll see if i can decompile it later and also check the internal files.

Reply:
There are some odd file formats in old phones.  I recently worked out the bizarre encoding used by the music files in Munkiki's Castle's.  It turned out to be a binary version of an encoding used to send monophonic ringtones over sms.  For all we know, it's just raw PCM data without the RIFF header.  

Oh, I tried a few versions of that game, and they all seem to be working correctly.  That was in an old version from 2018.  I haven't tried the latest version.  I have a baby that refuses to go to sleep and I don't have the patience at the moment for this.

Reply:
Heh, don't worry. Get some rest whenever you can, i'll be cleaning up some other things in the PRs for a while.

Reply:
Alright, decompiled the AC: Revelations jar, and the only code related to the media player in there is located in h.java:
![image](https://user-images.githubusercontent.com/52329050/173257612-57c85ffa-52f1-42b9-9a04-06e18ffe5034.png)

Pretty standard stuff considering how the media.Player class is implemented in FreeJ2ME, so maybe it is passing the wav file correctly, but the codec used is the actual problem. After extracting the audio files from the game into a folder, i proceeded to check what the properties of the wav files were, and turns out that the codec used is ADPCM IMA WAV:
![image](https://user-images.githubusercontent.com/52329050/173258170-913bcfcf-9056-4afa-8d6b-515c279754ca.png)
![image](https://user-images.githubusercontent.com/52329050/173258186-08579516-454f-40cb-903b-ed138db149f8.png)

Then i moved on to check if the codecs used by other games were different.

To do that, i opted to compare with a game that did load the wav files correctly (after that equalsIgnoreCase commit, because the game passes them as X-wav), i went on to unpack "Shadow of Egypt 3D". That one doesn't even encrypt the audio files:
![image](https://user-images.githubusercontent.com/52329050/173257942-1b0a3a18-35d3-47a5-83b8-1f1467da248d.png)

I would add the encryption to the list of factors why the wav playback didn't work, but i doubt that it is the reason since the games themselves pass those streams to the Player. Then, after loading the wav in the editor, it showed a different codec:
![image](https://user-images.githubusercontent.com/52329050/173258321-3a2b049f-caf1-4bfc-a75f-9bc43d454226.png)


So far, it seems to be a problem with the codec. At least it's the most likely one. Either way, i gotta find another game that fails to load a wav file to be sure it's the codec.

Reply:
Oh, it occurred to me that you can (with some temporary code) just dump whatever the game tries to load when it calls `createPlayer` to a file, rather than trying to figure out which of the needlessly obscured assets in the jar is the file you want.  Again, I'm happy to handle the analysis and conversion if you send over an adpcm or two.

Reply:
Ok. But first, i wanted to confirm that it was indeed a content type error, so i threw a few more exceptions into the try-catch of the wavPlayer costructor, and surely enough:
![image](https://user-images.githubusercontent.com/52329050/173275915-bb93f8b9-6552-47cf-a4cc-09d1e31b9190.png)

That "Stream of unsupported format" string is part of the message generated by UnsupportedAudioFileException, thrown by the getAudioInputStream method used by the wavStream in the player's constructor. 
https://docs.oracle.com/javase/8/docs/api/javax/sound/sampled/AudioSystem.html#getAudioInputStream-java.io.InputStream- 


Now we have a clear idea of what's going on. And what's more: that's also thrown by NitroStreet Racing 2 and Motocross Trial Extreme.

Now for the ADPCM file, here it is
[adpcmFile.zip](https://github.com/hex007/freej2me/files/8887786/adpcmFile.zip)

Reply:
Lucky.  It's a proper wav file after all.  It even has a RIFF header.  This means it's not a VOX file and is probably ADPCM wrapped in a wav.  Information about that is damn near impossible to find, but I did find this:

https://web.archive.org/web/20131231000305/http://support.microsoft.com/kb/89879

"ADPCM, commonly termed as a form of compression, is a more efficient way of storing waveforms than 16-bit or 8-bit PCM. It only uses 4 bits per sample, taking up a quarter of the disk space of 16-bit PCM. However, the sound quality is inferior. Because the Windows Sound System hardware only understands 8/16-bit PCM, the computer must compress and decompress the ADPCM into/from PCM, which requires CPU time. 22 kHz mono ADPCM can be decompressed real-time (that is, while playing) on a 386SX/16 megahertz CPU. Higher sampling rates (44 kHz) or stereo files will take too long for a 386SX/16 to decompress, which causes skipping in the audio. 11 kHz mono ADPCM can be compressed real-time on a 386SX/16 computer. To do ADPCM, the computer must have the Audio Compression Manager (ACM) installed.

ADPCM stores the value differences between two adjacent PCM samples and makes some assumptions that allow data reduction. Because of these assumptions, low frequencies are properly reproduced, but any high frequencies tend to get distorted. The distortion is easily audible in 11 kHz ADPCM files, but becomes more difficult to discern with higher sampling rates, and is virtually impossible to recognize with 44 kHz ADPCM files. "

Anyhow, here's what the file looks like:

At the very end, we have something that looks like metadata starting at 4668 (123Ch)  We should keep an eye out for it.

52 49 46 46 5A 12 00 00 -- "RIFF" 4698 ('RIFF'+ length of data to follow) matches the file size of 4706 
57 41 56 45 66 6D 74 20 -- "WAVEfmt " all standard stuff so far
14 00 00 00 -- Chunk size: 20 (14h)  -- This is a bit odd... PCM is usually 16
11 00 -- Format: 17 (PCM is 1)  This is how we'll detect that it's ADPCM 
01 00 -- Channels: 1 - It's mono, which isn't surprising in the least.
40 1F 00 00 -- Sample Rate: 8khz (8000, 1F40h)
A0 0F 00 00 -- Data Rate: 4kbps (4000, FA0h  bytes/second -- This is a very short clip)
00 01 -- Block Align: 256 (100h) (This is wrong.  Maybe it's encoding something the adjusted sample range? )
04 00 -- Bits per sample 4 (4h)  Thank goodness this matches the Microsoft KB article I found!  

This is followed by 16 bytes I don't fully understand.  Normally, I'd expect to see "data" here, but there is space left over in the current chuck, which is followed by a seemingly pointless "fact" chunk.  

02 00  -- 2 (a mystery)  These last two values must relate to the ADPCM encoding, I just don't know how.
F9 01 -- 505 (just as mysterious)
66 61 63 74 -- "fact" --This is a chunkid as it's exactly at the end of the previous chunk (the 14h from earlier)
04 00 00 00 -- 4 -- The chuck length
82 23 00 00 -- 9090 -- Also complete mystery. 

The "data" chuck follows, with a length of 4608 (1200h)  This should be the 4bps ADPCM encoded data.  The difference between samples is stored, not the actual sample, so I'm going to guess that this means those 4 bits are 2's compliment signed numbers, giving values ranging from 7 to -8.  I don't have any idea what the 'assumptions' are, but we should be able to get something passable with a naive decoding, we'll see.  Audacity seems to be able to play these and Windows has no problem reading the metadata.  It might be worth digging into Audacity's source to see if there's some helpful comments about the format.  

So what happens 4608 bytes later?  It turns out, that lands us directly on address 4606 (123Ch) ! Exactly where it looked like the metadata starts!  

We don't need to be able to decode the metadata, but just for fun:
4C 49 53 54 -- "LIST" this is the chunk id 
1E 00 00 00 -- 30 bytes, the chunk length
49 4E 46 4F 49 4E 41 4D -- "INFOINAM" the key in a list of key-value pairs, presumably fixed size
12 00 00 00 -- 18, the length of the string that follows
73 66 78 5F 62 6F 73 73 5F 6C 69 67 68 74 6E 69 6E 67 -- 18 byte string, "sfx_boss_lightning"

Okay, so what does all this mean?  Well, it does look like we can these files very easily.  We can already read all of the relevant parts and we have a good idea how to read the ADPCM data (The compression scheme itself is very simple.)   Generating a proper wav stream from this should be very easy.  I even have some code from another personal project that I can repurpose.  

The only annoying part is that we'll need to make a copy of every stream we suspect could be ADPCM.  With luck, that'll just be type "audio/wav" but you never know.

Reply:
It seems audacity relies on ffmpeg to handle formats like WAV with ADPCM, and fortunately, ffmpeg is not only open-source as well, but it also has a specific file containing the decode data, with reference links as well:
https://github.com/FFmpeg/FFmpeg/blob/6ed0d95fa207f587b3542f3e8349932118ad8d5d/libavcodec/adpcm.c

I'll be looking at it later, i'm a bit busy currently.

Reply:
I've found a few others as well, along with some lecture notes that explain the encoding/decoding process.  I spent some time comparing different  implementations last night, and found some answers, like what the 256, 2, and 505 are for and how to construct the step table (it's just an exponential) and the other tables.  My assumption about the 4-bit values wasn't too far off, bit 3 is the sign bit, the others change the delta size by multiplying the step value by some amount.  This explains the adaptive part in ADPCM.  I'll try to play with this more on the weekend.

Copying this here so I don't lose it, from : https://github.com/dbry/adpcm-xq/blob/master/adpcm-lib.c

"This module encodes and decodes 4-bit ADPCM (DVI/IMA varient). ADPCM data is divided into independently decodable blocks that can be relatively small. **The most common configuration is to store 505 samples into a 256 byte block**, although other sizes are permitted as long as the number of samples is one greater than a multiple of 8. When multiple channels are present, they are interleaved in the data with an 8-sample interval. "